### DISCUSSION TITLE
Support use of ScrollRestoration that does not automatically scroll to hash elements for specific `navigate` calls

### DISCUSSION THREAD

**Discussion by sjdemartini** [remix-run/react-router#10038] [Proposals]
As originally described in https://github.com/remix-run/react-router/pull/10034, the `useScrollRestoration` always scrolls to the hash element (if found) when navigating to a route containing a hash. See here: https://github.com/remix-run/react-router/blob/6120207a328740a226c462d87be479b7c26b9574/packages/react-router-dom/index.tsx#L1207-L1214

This behavior is regardless of the `preventScrollReset` value. That means that it's currently impossible to use `useScrollRestoration`/`<ScrollRestoration />` and control the scrolling behavior when using `navigate` or `Link`s that point to routes with hashes. Apps are forced to have the app immediately jump to the hash element with no workaround. 

A couple scenarios where you would *not* want to immediately jump to a hash:
1. When an app doesn't want the scroll position to change at all in response to a specific `Link`/`navigate` route that includes a hash, but does want the URL to be updated. Tabbed interfaces are one example. This is essentially the same behavior as what `preventScrollReset` provides when a hash is not present, in that scroll position is *maintained* when navigating to the new route (no automatic scroll to any location).
2. When an app wants control over the scroll behavior for a specific action. For instance, someone clicks a button that should link to a given hash element, so we want to update the router state history (using navigate or Link), but want to smooth-scroll to that element ourselves and don't want `ScrollRestoration` to immediately jump for us. (Or maybe you need transitions, etc.)

One suggested workaround from the original PR https://github.com/remix-run/react-router/pull/10034#issuecomment-1414472992 is to use `<ScrollRestoration getKey>`. However, this is different in behavior than what `preventScrollReset` provides for non-hash routes and is more limiting, since `preventScrollReset` allows you to avoid changing the scroll position during _specific navigation events_ added to the stack, rather than storing/restoring scroll position based on the `location` alone. (Changing `getKey` has consequences for revisiting pages, making it quite different from `preventScrollReset`.)

As such, it seems useful for `react-router` to provide the automated scroll restoration like it does today (so position on a page is preserved when using back/forward, and goes to the top or a hash when you want it to), while still allowing users to "opt out" of scroll-to functionality for specific `navigate` calls to routes, _even if the route contains a hash_. I think having the automatic "scroll to hash" behavior only run when `preventScrollReset !== true` is sufficient and intuitive to achieve this, since that `preventScrollReset` option is currently ignored entirely when a hash is present, and it would give users the same level of control as it does for non-hash links. Also open to other alternative APIs to achieve this.

**Comment by lovelidevs** [comment]
I am very interested in having smooth scroll to hash elements. It would be great to just have a new prop on the `Link` element, such as `scroll`, to handle this. Similar to how `react-router-hash-link` did it:
https://github.com/rafgraph/react-router-hash-link#smooth-boolean

I currently have this working by adding `scroll-smooth` (tailwind) to my `html` element in `root.tsx`, but there are downsides:
https://css-tricks.com/downsides-of-smooth-scrolling/

**Comment by penx** [comment]
I've been trying to implement (2) above for a little while today and found this discussion.

Our use case is to:

- highlight certain sections of the page based on the hash (within ClientOnly)
- the browser URL should be sharable/bookmarkable as-is, and contain a hash with real element ids
- use scrollIntoView when navigating to these elements using anchors within the page, but persist hash in the URL with the  correct element id

It would be great to have a pattern available from react-router/remix to do this, I don't see a solution in the comments in https://github.com/remix-run/react-router/pull/10034#issuecomment-1414472992

I could use the browser history API directly:

```tsx
function Bookmark({ id, children }: { id: string; children: React.ReactNode }) {
  return (
    <a
      id={`${id}`}
      href={`#${id}`}
      onClick={(event) => {
        history.pushState({}, "", `#${id}`);
        document
          .getElementById(id)
          ?.scrollIntoView({ behavior: "smooth", block: "nearest", inline: "nearest" });
        event.preventDefault();
      }}
    >
      {children}
    </a>
  );
}
```

This works but doesn't cause a rerender if anything is monitoring `useLocation().hash`, and using the browser history API directly is generally is not recommended with react-router.

I could use `useNavigate`:

```tsx
function Bookmark({ id, children }: { id: string; children: React.ReactNode }) {
  const navigate = useNavigate();
  return (
    <a
      id={`${id}`}
      href={`#${id}`}
      onClick={(event) => {
        navigate(
          {
            hash: id,
          },
          { replace: true, preventScrollReset: true },
        );
        document
          .getElementById(id)
          ?.scrollIntoView({ behavior: "smooth", block: "nearest", inline: "nearest" });
        event.preventDefault();
      }}
    >
      {children}
    </a>
  );
}
```

...but this causes the browser to scroll directly to the element mid animation.

@brophdawg11 Perhaps a scrollIntoView configuration option could be added to NavigateOptions?

```ts
navigate(
  {
    hash: id,
  },
  { scrollIntoView: true },
);
```

**Comment by fr1n63** [comment]
Experiencing the same issue with a near identical pattern for highlighting elements on a page - additional comment is that `ScrollRestoration` triggers even if the hash doesn't change - so `navigate({hash: '#${id}'}, {replace: true})` when `id === hash` still triggers a scroll, whilst `const {hash} = useLocation();` won't have changed

There's a distinction to be made between a change in the url that has triggered an actual route change ie: updated outlets - and something that is just changing the url, which I'd prefer to retain control over.
